Clear/EVMState.lean

```diff
@@ -28,19 +28,110 @@ def ByteArray.byteArrayToUInt256 (μ₀ : UInt256) (size : ℕ) (Id : ByteArray)
   let r : (ℕ × UInt256) := Array.foldl step ((size - 1) * 8, 0) arr1
   r.2
 
-
 namespace Clear
 
--- 2^160 https://www.wolframalpha.com/input?i=2%5E160
-def Address.size : Nat := 1461501637330902918203684832716283019655932542976
-
+def Address.size : Nat := 2 ^ 160
 abbrev Address : Type := Fin Address.size
 
 instance : Inhabited Address := ⟨Fin.ofNat 0⟩
+instance : NeZero Address.size := ⟨by decide⟩
+
+namespace Address
+
+def top : ℕ := (⊤ : Address).val
+
+lemma top_def : top = 2 ^ 160 - 1 := by
+  unfold top
+  rw [Fin.top_eq_last]
+  simp
+
+lemma top_def' : top = 1461501637330902918203684832716283019655932542975 := by
+  rw [top_def]; simp
+
+lemma size_def : size = 1461501637330902918203684832716283019655932542976 := by
+  unfold size; simp
+
+def ofNat (n : ℕ) : Address := Fin.ofNat n
+def ofUInt256 (u : UInt256) : Address := Fin.ofNat (u.val % size)
+
+instance {n : Nat} : OfNat Address n := ⟨ofNat n⟩
+instance : NeZero top := ⟨by decide⟩
+
+lemma zero_lt_top  : 0 < top := by decide
+lemma zero_le_top  : 0 ≤ top := le_of_lt zero_lt_top
+
+lemma zero_lt_size : 0 < size := by decide
+lemma zero_le_size : 0 ≤ size := le_of_lt zero_lt_size
+
+lemma one_lt_top  : 1 < top := by decide
+lemma one_le_top  : 1 ≤ top := le_of_lt one_lt_top
+
+lemma one_lt_size : 1 < size := by decide
+lemma one_le_size : 1 ≤ size := le_of_lt one_lt_size
+
+lemma top_eq_pred_size : top = size - 1 := by
+  unfold size top
+  rw [Fin.top_eq_last]
+  simp
+
+lemma top_eq_pred_size_ofUInt256 : top.toUInt256 = size.toUInt256 - 1 := by
+  unfold size top
+  rw [Fin.top_eq_last]
+  simp
+  decide
+
+lemma succ_top_eq_size : top + 1 = size := by
+  rw [top_eq_pred_size, Nat.sub_add_cancel (le_of_lt one_lt_size)]
+
+lemma top_lt_size : top < size := by
+  rw [← succ_top_eq_size]; simp
+
+lemma top_le_size : top ≤ size := le_of_lt top_lt_size
+
+lemma size_lt_UInt256_size : size < UInt256.size := by decide
+lemma size_le_UInt256_size : size ≤ UInt256.size := le_of_lt size_lt_UInt256_size
+
+lemma top_lt_UInt256_size : top < UInt256.size := by decide
+lemma top_le_UInt256_size : top ≤ UInt256.size := le_of_lt top_lt_UInt256_size
+
+lemma ofUInt256_lt_UInt256_size {u : UInt256} : ↑(ofUInt256 u) < UInt256.size := by
```

Comment:
Shorted version that doesn't use open `simp`s:
```
lemma ofUInt256_lt_UInt256_size {u : UInt256} : ↑(ofUInt256 u) < UInt256.size := by
  unfold ofUInt256 Fin.ofNat
  trans (2 ^ 160)
  · simp only [UInt256.size, Nat.reducePow, Nat.reduceAdd, gt_iff_lt, Nat.ofNat_pos, Nat.mod_lt]
  · simp only [Nat.reducePow, UInt256.size, Nat.reduceLT]
  ```

Reply:
In general, we have a strong preference for only using open `simp`s as a closing tactic.